fem/estimators.cpp

```diff
@@ -48,6 +48,283 @@ void L2ZienkiewiczZhuEstimator::ComputeEstimates()
    current_sequence = solution->FESpace()->GetMesh()->GetSequence();
 }
 
+void KellyErrorEstimator::ComputeEstimates()
+{
+   // Remarks:
+   // For some context you may have to consult the documentation of
+   // the FaceInfo class [1]. Also, the FaceElementTransformations
+   // documentation [2] may be helpful to grasp what is going on. Note
+   // that the FaceElementTransformations also works in the non-
+   // conforming case to transfer the gauss points from the slave to
+   // the master element.
+   // [1]
+   // https://github.com/mfem/mfem/blob/02d0bfe9c18ce049c3c93a6a4208080fcfc96991/mesh/mesh.hpp#L94
+   // [2]
+   // https://github.com/mfem/mfem/blob/02d0bfe9c18ce049c3c93a6a4208080fcfc96991/fem/eltrans.hpp#L435
+   flux_space->Update(false);
+
+   auto xfes      = solution->ParFESpace();
+   MFEM_ASSERT(xfes->GetVDim() == 1, "Estimation for vector-valued problems not implemented yet.");
+   auto pmesh     = xfes->GetParMesh();
+
+   this->error_estimates.SetSize(xfes->GetNE());
+   this->error_estimates = 0.0;
+
+   // 1. Compute fluxes in discontinuous space
+   ParGridFunction flux(flux_space);
+   flux = 0.0;
+
+   Array<int> xdofs, fdofs;
+   Vector el_x, el_f;
+   for (int e = 0; e < xfes->GetNE(); e++)
+   {
+      auto attr = xfes->GetAttribute(e);
+      if (attributes.Size() && attributes.Find(attr) == -1)
+      {
+         continue;
+      }
+
+      xfes->GetElementVDofs(e, xdofs);
+      solution->GetSubVector(xdofs, el_x);
+
+      ElementTransformation* Transf = xfes->GetElementTransformation(e);
+      flux_integrator->ComputeElementFlux(*xfes->GetFE(e), *Transf, el_x,
+                                          *flux_space->GetFE(e), el_f, true);
+
+      flux_space->GetElementVDofs(e, fdofs);
+      flux.AddElementVector(fdofs, el_f);
+   }
+
+   // 2. Add error contribution from local interior faces
+   auto int_rules = IntegrationRules();
```

Comment:
No need to instantiate this container -- just use the global one: https://github.com/mfem/mfem/blob/4645f97f0f53a37c3245f7efd777ab0640d728ef/fem/intrules.cpp#L877